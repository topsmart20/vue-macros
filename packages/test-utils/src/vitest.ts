import path from 'node:path'
import glob, { type Options as GlobOptions } from 'fast-glob'
import { describe, expect, test } from 'vitest'
import { normalizePath } from '@rollup/pluginutils'

interface Options {
  params?: [name: string, values?: any[]][]
  promise?: boolean
}

const SKIP_VUE2 = !!process.env.SKIP_VUE2

export async function testFixtures(
  globs: string | string[],
  exec: (args: Record<string, any>, id: string) => any,
  options?: GlobOptions & Options,
): Promise<void>
export async function testFixtures(
  files: Record<string, string>,
  exec: (args: Record<string, any>, id: string, code: string) => any,
  options?: Options,
): Promise<void>
export async function testFixtures(
  globsOrFiles: string | string[] | Record<string, string>,
  cb: (args: Record<string, any>, id: string, code: string) => any,
  { params, promise, ...globOptions }: GlobOptions & Options = {},
) {
  let files: Record<string, string | undefined>

  if (typeof globsOrFiles === 'string' || Array.isArray(globsOrFiles)) {
    files = Object.fromEntries(
      (await glob(globsOrFiles, globOptions)).map((file) => [file, undefined]),
    )
  } else {
    files = globsOrFiles
  }

  for (const [id, code] of Object.entries(files)) {
    makeTests(id, code, [[normalizePath(id)], ...(params || [])])()
  }

  function getName(name: string, value: any) {
    return value !== undefined ? `${name} = ${String(value)}` : name
  }

  function makeTests(
    id: string,
    code: string | undefined,
    params: NonNullable<Options['params']>,
    args: Record<string, any> = {},
  ) {
    const [currParams, ...restParams] = params
    const [name, values = [undefined]] = currParams
    if (restParams.length > 0) {
      return () => {
        for (const value of values) {
          const currArgs = { ...args, [name]: value }

          describe(
            getName(name, value),
            makeTests(id, code, restParams, currArgs),
          )
        }
      }
    } else {
      return () => {
        for (const value of values) {
          const testName = getName(name, value)
          const isSkip = testName.includes('vue2') && SKIP_VUE2

          test.skipIf(isSkip)(testName, async () => {
            const currArgs = { ...args, [name]: value }
            const execute = () =>
              cb(currArgs, path.resolve(globOptions.cwd || '/', id), code!)
            if (id.includes('error')) {
              if (promise) {
                await expect(execute()).rejects.toThrowErrorMatchingSnapshot()
              } else {
                expect(execute).toThrowErrorMatchingSnapshot()
              }
            } else if (promise) {
              await expect(
                (execute() as Promise<any>).catch((error) => {
                  console.warn(error)
                  return Promise.reject(error)
                }),
              ).resolves.toMatchSnapshot()
            } else {
              expect(execute()).toMatchSnapshot()
            }
          })
        }
      }
    }
  }
}
