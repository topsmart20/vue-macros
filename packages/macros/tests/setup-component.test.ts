/// <reference types="vite/client" />
import { resolve } from 'node:path'
import { describe, expect, it } from 'vitest'
import glob from 'fast-glob'
import esbuild from 'rollup-plugin-esbuild'
import Vue from 'unplugin-vue/vite'
import VueJsx from '@vitejs/plugin-vue-jsx'
import VueMacros from '../src/rollup'
import { getCode } from './_utils'

describe('setup-component', async () => {
  const root = resolve(__dirname, '..')
  const files = await glob('tests/fixtures/setup-component/*.{vue,[jt]s?(x)}', {
    cwd: root,
    onlyFiles: true,
  })

  for (const file of files) {
    it(file.replace(/\\/g, '/'), async () => {
      const filepath = resolve(root, file)
      const version = filepath.includes('vue2') ? 2 : 3

      const unpluginCode = await getCode(filepath, [
        VueMacros({ version }),
        Vue() as any,
        VueJsx(),
        esbuild(),
      ])
      expect(unpluginCode).toMatchSnapshot()
    })
  }
})
