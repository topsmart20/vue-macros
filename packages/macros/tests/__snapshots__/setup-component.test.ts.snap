// Vitest Snapshot v1

exports[`setup-component > tests/fixtures/setup-component/async.tsx 1`] = `
"import { defineComponent, withAsyncContext } from 'vue';

var _sfc_main = /* @__PURE__ */ defineComponent({
  __name: \\"0\\",
  props: {
    foo: {
      type: String,
      required: true
    }
  },
  async setup(__props) {
    let __temp, __restore;
    [__temp, __restore] = withAsyncContext(() => foo()), await __temp, __restore();
    return () => {
    };
  }
});

var _export_sfc = (sfc, props) => {
  const target = sfc.__vccOpts || sfc;
  for (const [key, val] of props) {
    target[key] = val;
  }
  return target;
};

var setupComponent_0 = /*#__PURE__*/_export_sfc(_sfc_main, [['__file',\\"#ID#\\"]]);

const App = setupComponent_0;

export { App };
"
`;

exports[`setup-component > tests/fixtures/setup-component/basic.ts 1`] = `
"import { defineComponent } from 'vue';

const __default__ = defineComponent({
  name: \\"App\\"
});
var _sfc_main = /* @__PURE__ */ defineComponent({
  ...__default__,
  props: {
    foo: { type: String, required: true }
  },
  emits: [\\"change\\"],
  setup(__props) {
    return () => {
    };
  }
});

var _export_sfc = (sfc, props) => {
  const target = sfc.__vccOpts || sfc;
  for (const [key, val] of props) {
    target[key] = val;
  }
  return target;
};

var setupComponent_0 = /* @__PURE__ */ _export_sfc(_sfc_main, [[\\"__file\\", \\"#ID#\\"]]);

const App = setupComponent_0;

export { App };
"
`;

exports[`setup-component > tests/fixtures/setup-component/render.tsx 1`] = `
"import { defineComponent, createVNode } from 'vue';

const __default__ = defineComponent({
  name: \\"RenderApp\\",
  render() {
    return createVNode(\\"div\\", null, null);
  }
});
var _sfc_main = /* @__PURE__ */ defineComponent({
  ...__default__,
  props: {
    foo: {
      type: String,
      required: true
    }
  },
  setup(__props) {
    return {};
  }
});

var _export_sfc = (sfc, props) => {
  const target = sfc.__vccOpts || sfc;
  for (const [key, val] of props) {
    target[key] = val;
  }
  return target;
};

var setupComponent_0 = /*#__PURE__*/_export_sfc(_sfc_main, [['__file',\\"#ID#\\"]]);

const App = setupComponent_0;

export { App };
"
`;
